Тестовые задания на уровень осведомленности.

1: Напишите код на Python, после выполнения которого из словаря будет получено число 80. Запишите ответ в одну строку без пробелов.
Код должен начинаться с имени словаря. НЕ используйте функцию print и НЕ присваивайте результат работы кода какой-либо дополнительной переменной

In [1]:
my_dict = {'красный':[{'помидор': 115}, {'малина': 10}], 

           'желтый':[{'лимон': 80}, {'абрикос': 32}], 

           'зеленый':[{'яблоко': 160}, {'арбуз': 1400}]}

In [10]:
my_dict['желтый'][0]['лимон']

80

In [7]:
for shortlist in my_dict.values(): 
    for thing in shortlist:
        for price in thing.values():
            if price==80:
                print(price)

80


2: У нас есть код, который преобразует строку стихотворения Пушкина в список.   
Напишите код на Python, который выдает список слов, соответствующий второй строке стихотворения (еще ты дремлешь друг прелестный).

In [11]:
pushkin_list = "Мороз и солнце день чудесный еще ты дремлешь друг прелестный пора красавица проснись открой сомкнутой негой взоры".split(" ") 

In [13]:
new_list=[]
check_list='еще ты дремлешь друг прелестный'.split(" ")
for word in pushkin_list:
    if word in check_list:
        new_list.append(word)
print(new_list)

['еще', 'ты', 'дремлешь', 'друг', 'прелестный']


In [14]:
pushkin_list[5:10]

['еще', 'ты', 'дремлешь', 'друг', 'прелестный']

3: Напишите функцию count_a_words(s), которая возвращает список слов, содержащих русскую букву а (большую или маленькую) в строке s.
Обратите внимание, что функция должна принимать в качестве аргумента переменную типа str, и в результате работы возвращать список (list).

In [30]:
s=input('?')
new_list=[]
def count_a_words(s):
    for word in s.split(" "):
        for letter in word:
            if letter=='а' or letter=='А':
                new_list.append(word)
    return new_list
d=count_a_words(s)
print(d)                          #не принял валидатор ответ

?да но нет но можна
да
можна
['да', 'можна']


Начальные задания:

In [16]:
a=5
b=6
print("a=", a, "b=", b,  sep='***', end='')     #печатаем строку с разделителями

a=***5***b=***6

In [ ]:
sourceFile=open('python.txt','w')              #создаем и открываем текстовый файл, записываем в него инфо, закрываем
print('this is cool!', file=sourceFile)
sourceFile.close()

In [17]:
for i in range(1,10):                         #выводим таблицу умножения
    for j in range (1,10):
        m=i*j
        print(m, end='\t')
    print('')   
       

1	2	3	4	5	6	7	8	9	
2	4	6	8	10	12	14	16	18	
3	6	9	12	15	18	21	24	27	
4	8	12	16	20	24	28	32	36	
5	10	15	20	25	30	35	40	45	
6	12	18	24	30	36	42	48	54	
7	14	21	28	35	42	49	56	63	
8	16	24	32	40	48	56	64	72	
9	18	27	36	45	54	63	72	81	


In [19]:
'hey'+'!'                                     #можно просто писать выражение, будет выведен результат

'hey!'

In [20]:
(1267-20*(5**3/2)**2)/(8**5*36**(1/2))        #можно просто писать выражение, будет выведен результат

-0.3909200032552083

In [21]:
word='Очаровательно'
print('Длина слова ',word,len(word),' букв')  #оператор len определяет длину строки

Длина слова  Очаровательно 13  букв


In [22]:
word=input('Введите любое слово:')
if len(word)>7:
    print('это длинное слово')
else:
    print('это короткое слово')

Введите любое слово:строка
это короткое слово


In [23]:
word=input('Введите любое слово:')
letter=input('Какую букву проверим?')
if letter in word:                                      #оператор in проверяет наличие символа в строке
    print('Выбранная буква есть в введённом слове')
else:
    print('Выбранной буквы нет в введённом слове')

Введите любое слово:рука
Какую букву проверим?е
Выбранной буквы нет в введённом слове


In [24]:
number=int(input('Введите любое положительное число:'))
b=number**(1/2)
a=int(b)                                                      #оператор int() делает число(строку) целым
c=b-a
if c==0: 
    print(b)
else: 
    print('Извлечь корень без остатка невозможно.')

Введите любое положительное число:7
Извлечь корень без остатка невозможно.


In [25]:
weight=float(input('Введите ваш вес:'))                    #оператор float() делает число(строку) вещественным
height=float(input('Введите ваш рост:'))
ИМТ=weight/height**2
if ИМТ<18.5: 
    print("Недостаточная масса тела")
elif ИМТ>24.99: 
    print("Избыточная масса тела")
else: 
    print('Норма')  

Введите ваш вес:68
Введите ваш рост:1.85
Норма


In [26]:
hour=int(input('Введите час:'))                          #логические операторы and и or
minute=int(input('Введите минуты:'))
if hour==10 and minute<30 or hour>=12 and hour<=13 and minute<40 or hour>=15 and hour<18 or hour==19 and minute>30:
    print('Преподаватель свободен.')
else:
    print('Преподаватель занят.')

Введите час:12
Введите минуты:45
Преподаватель занят.


In [27]:
n = int(input('n = '))
for j in range(1, n, 2):                            #оператор range(А) берет числа от 0 до А, не включая А. 
    print(j, end=' ')
else:
    if n % 2:
        print(n)

n = 9
1 3 5 7 9


In [28]:
sequence=[1,2,7,19]                                    #список
for idx, item in enumerate(sequence, 1):               #оператор enumerate выдает пару номер элемента+элемент
    print(idx, item)
   

1 1
2 2
3 7
4 19


In [29]:
f = None                                             #числа Фибоначчи
f1, f2 = 1, 1
n = int(input('n = '))
print('1 =>', f1)
print('2 =>', f2)
for j in range(3, n + 1):
    f  = f1 + f2
    f1, f2 = f2, f
    print(j, '=>', f)

n = 4
1 => 1
2 => 1
3 => 2
4 => 3


1.19 Общий синтаксис цикла с параметром  

Задание 1:
Друзья придумали свой шифр: исходные слова заменять на последовательность нечётных символов в этих словах. Например, 'python' в их шифровке будет выглядеть как 'pto'. Напишите программу, которая помогает зашифровывать слова по этой схеме: принимает на вход слово (word) и последовательно выводит каждый нечётный символ в нём. Для проверки используйте word = 'подшипник'

In [1]:
word=input('Введите слово:')
for item, letter in enumerate(word,1):
    if item%2==1:
        print(letter, end='')

Введите слово:подшипник
пдинк

То, что приняла проверка. Вывод в столбик.

In [2]:
word=input('Введите слово:')
for item, letter in enumerate(word,1):
   if item%2==1:
        print(letter)

Введите слово:подшипник
пдинк

Задание 2: 
Вы решили написать программу, которая выявляет високосные года в диапазоне лет. Пусть она принимает на вход два числа: нижняя и верхнюю границу годов (year_1, year_2) (например, 2013 и 2020) и для каждого введённого года выводит информацию: "... год високосный" или "... год невисокосный". Для проверки используйте year_1 = 2013, year_2 = 2020

In [9]:
year_1=int(input("введите первый год: \n"))
year_2=int(input("введите второй год: \n"))
for i in range (year_1,year_2+1):
    if i%4==0 and i%100!=0 or i%100==0 and i%400!=0:
        print(i, "- это високосный год")
    else:
        print(i, "- это невисокосный год")

введите первый год: 
2013
введите второй год: 
2020
2013 - это невисокосный год
2014 - это невисокосный год
2015 - это невисокосный год
2016 - это високосный год
2017 - это невисокосный год
2018 - это невисокосный год
2019 - это невисокосный год
2020 - это високосный год


Задание 3:   
Напишите программу, которая принимает на вход слово (word) и последовательно выводит все русские гласные из этого слова. Для проверки используйте word = 'Зимушка-зима'

In [13]:
word=input('Введите слово:').lower()
for item, letter in enumerate(word,1):
    if letter=='а' or letter=='и' or letter=='е' or letter=='ё' or letter=='о' or letter=='у' or letter=='ы' or letter=='э' or letter=='ю'  or letter=='я':
        print(letter, end='')

Введите слово:ЗИмушка-Зима
иуаиа

То, что приняла проверка. Вывод в столбик:

In [ ]:
word=input('Введите слово:').lower()
for item, letter in enumerate(word,1):
    if letter=='а' or letter=='и' or letter=='е' or letter=='ё' or letter=='о' or letter=='у' or letter=='ы' or letter=='э' or letter=='ю'  or letter=='я':
        print(letter)

Задание 4:
Попросите пользователя ввести своё имя (name) и с помощью цикла выведите каждую букву этого имени на экран, сопроводив порядковым номером с помощью фразы: "Буква ... в этом имени - ..." (например, "Буква 1 в этом имени - С"). Для проверки используйте name = 'Сергей'

In [15]:
name=input('Введите ваше имя:')
for i, letter in enumerate(name,1):
    print( "Буква {} в этом имени - {}".format(i,letter))

Введите ваше имя:Вика
Буква 1 в этом имени - В
Буква 2 в этом имени - и
Буква 3 в этом имени - к
Буква 4 в этом имени - а


Задание 5:
Напишите программу, которая последовательно выводит нечётные числа, которые делятся на 3 или 5, от первого введённого числа (num_1) до второго введённого числа (num_2) включительно. Если второе число меньше первого или первое число меньше нуля, выведите "Введён неверный диапазон чисел".
Для проверки используйте num_1 = 25, num_2 = 45

In [19]:
num_1=int(input('Введите первое число:'))
num_2=int(input('Введите второе число:'))
if num_1>0 and num_1<num_2:
    for j in range(num_1,num_2+1):
        if j%2==1 and (j%3==0 or j%5==0):
            print(j)
else:
    print('Введён неверный диапазон чисел')
        

Введите первое число:5
Введите второе число:2
Введён неверный диапазон чисел


1.20 Вычисления в циклах

Задание 1 
Выведите таблицу квадратов и кубов для чисел от 1 до 10. (Каждая новая строчка распечатывается через print() и имеет формат '2 4 8', то есть число, затем его квадрат, а затем его куб)

In [22]:
for i in range(1,11):
    print(i, i**2, i**3)

1 1 1
2 4 8
3 9 27
4 16 64
5 25 125
6 36 216
7 49 343
8 64 512
9 81 729
10 100 1000


Задание 2 
Напишите программу, которая получает на вход число (number) и последовательно в порядке возрастания выводит все его делители на экран (само число тоже является своим делителем). 

In [24]:
number=int(input('Введите число:'))
for n in range(1,number+1):
        if number%n==0:
            print(n)

Введите число:10
1
2
5
10


Задание 3 
Написать программу, которая ищет все общие (не равные 1) делители двух чисел (num_1 и num_2), введённых пользователем, и последовательно в порядке возрастания выводит их на экран. Если общих делителей нет, не выводит на экран ничего. 
Для проверки используйте num_1 = 12 и num_2 = 38

In [28]:
num_1=int(input('Введите первое число:'))
num_2=int(input('Введите второе число:'))
if num_1<num_2:
    for c in range(2,num_2+1):
        if num_1%c==0 and num_2%c==0:
            print(c)
else:
     for c in range(2,num_1+1):
        if num_1%c==0 and num_2%c==0:
            print(c)

Введите первое число:10
Введите второе число:20
2
5
10


In [46]:
num_1=int(input('Введите первое число:'))   #так тоже работает... по сути, этот вариант лучше тем, что проще
num_2=int(input('Введите второе число:'))
for c in range(2,num_2+1):
        if num_1%c==0 and num_2%c==0:
            print(c)


Введите первое число:1812
Введите второе число:2500
2
4


Задание 4 
Написать программу, которая на вход получает натуральное цисло n и выводит n-е число ряда Фибоначчи на экран. 
Для проверки используйте n = 10

In [35]:
n=int(input('Введите число:'))
Fibonacci_n=(((1+5**(1/2))/2)**n-((1-5**(1/2))/2)**n)/(5**(1/2))    #формула Бине
print(int(Fibonacci_n))

Введите число:3
2


In [36]:
n=int(input('Введите число:'))
Fibonacci_0=0
Fibonacci_1=1
for i in range(2,n+1):
    Fibonacci_n=Fibonacci_1+Fibonacci_0
    Fibonacci_0=Fibonacci_1
    Fibonacci_1=Fibonacci_n
print(Fibonacci_n)

Введите число:3
2


Задание 5  
Напишите программу, которая выводит таблицу умножения от 1 до 10 для выбранного целого числа (number). Если number = 15, то строка таблицы имеет вид "15 X 3 = 45". 
Для проверки используйте number = 15

In [42]:
number=int(input('Введите число:'))
for i in range(1,11):
    result=number*i
    print(number, "X", i, "=", result)

Введите число:15
15 X 1 = 15
15 X 2 = 30
15 X 3 = 45
15 X 4 = 60
15 X 5 = 75
15 X 6 = 90
15 X 7 = 105
15 X 8 = 120
15 X 9 = 135
15 X 10 = 150


1.21 Управление выполнением цикла

Задание 1 
Напишите программу, которая ищет наибольший (не равный 1) общий делитель двух чисел, введённых пользователем (num_1, num_2), и выводит его на экран. Если общих делителей нет, выводит на экран "Общих делителей не найдено". 
Для проверки используйте num_1 = 1812, num_2 = 2500
PS: Если использовать break в цикле for - else, то при прерывании цикла код после else выполняться не будет.

In [52]:
num_1=int(input('Введите первое число:'))
num_2=int(input('Введите второе число:'))
maxim=0
for c in range(1,num_2+1):                            #не поняла, как делать с break или continue
    if num_1%c==0 and num_2%c==0:
        maxim=c
if maxim>1:
    print(maxim)
else:
    print('Общих делителей не найдено')

Введите первое число:5
Введите второе число:10
5


Задание 2 
Напишите программу с использованием оператора continue, которая принимает на вход строку (string), после чего выводит её посимвольно, но без букв а, б, в. Пример: ввод "арбуз", вывод "руз". 

Для проверки используйте string = 'абстракция'

PS: Чтобы выводить побуквенно в одну строчку, в фукнции print предусмотрен параметр end. Если написать print('Hallo!', end=''), следующий print продолжит вывод не с новой строки.

In [61]:
string=input('Введите слово:')
for idx, letter in enumerate(string):
    if letter=='а' or letter=='б' or letter=='в':
        continue
    print(letter, end='')

Введите слово:абстракция
стркция

Задание 3 
Напишите программу, которая анализирует ответы пользователя на вопрос "Любите ли вы Python?". Если ответ пользователя (string) - "Да", то программа отвечает: "Это отлично!", иначе программа отвечает: "Увы, это неправильный ответ". Программа должна "спросить" пользователя максимум 5 раз. При пяти ответах "Нет" выводит: "Это безнадёжно!". 
Для проверки используйте string = 'Нет', а на каждой итерации цикла не меняйте значение переменной.
Команду, выводящую на экран вопрос: "Любите ли вы Python?", с код включать НЕ нужно. 

In [5]:
c=0
for i in range(5):
    answer=input('Любите ли вы Python:').lower()
    if answer=='да':
        print('Это отлично!')
        break
    else:
        print('Увы, это неправильный ответ')
        c=c+1
        if c==5:
            print('Это безнадёжно!')        

Любите ли вы Python:да
Это отлично!


In [11]:
c=0
for i in range(5):
    answer=input('Любите ли вы Python:').lower()
    if answer=='нет':
        c=c+1
        if c==5:
            print('Это безнадёжно!')
            continue
        print('Увы, это неправильный ответ')     
    else:
        print('Это отлично!')
        break      

Любите ли вы Python:нет
Увы, это неправильный ответ
Любите ли вы Python:нет
Увы, это неправильный ответ
Любите ли вы Python:нет
Увы, это неправильный ответ
Любите ли вы Python:нет
Увы, это неправильный ответ
Любите ли вы Python:нет
Это безнадёжно!
Любите ли вы Python:да
Это отлично!


In [14]:
for i in range(5):                              #то, что приняла проверка
    string='Нет'
    if string=='Нет':
        print('Увы, это неправильный ответ')  
        if i==4:
            print('Это безнадёжно!')
            continue
    else:
        print('Это отлично!')
        break      

Увы, это неправильный ответ
Увы, это неправильный ответ
Увы, это неправильный ответ
Увы, это неправильный ответ
Увы, это неправильный ответ
Это безнадёжно!


Задание 4 
Напишите программу, которая проверяет, является ли введённое число (number) простым. Если число простое, то программа выводит "Простое", иначе "Не является простым". 

Для проверки используйте number = 173

In [24]:
number=int(input('Введите число:'))
for i in range(2,number):
    if number%i==0:
        print('Не является простым')
        break
else:
    print('Простое')

Введите число:173
Простое


Задание 5 
Напишите программу, которая проверяет введённую пользователем строку (string) на соответствие правилу: у строки нет ни одной гласной русской маленькой буквы на каждой чётной позиции. Пример подходящей строки: "пока!" (в Python первый символ является нулевым, поэтому мы считаем первый символ чётным). Если строка подходит, программа выводит на экран "Какая хорошая строка!", а иначе "Строка мне не нравится!". 

Для проверки используйте string = 'прелестная строка'

In [28]:
string=input('Введите строку:').lower()
for item, letter in enumerate(string, 0):
    if item%2==0 and (letter=='а' or letter=='и' or letter=='е' or letter=='ё' or letter=='о' or letter=='у' or letter=='ы' or letter=='э' or letter=='ю'  or letter=='я'):
        print("Строка мне не нравится!")
        break
else:
    print("Какая хорошая строка!")    

Введите строку:пока
Какая хорошая строка!


1.22 Цикл while

Задание 1 
Клиент банка открыл вклад под 10% годовых и положил на этот счёт Y рублей. Напишите программу, позволяющую определить, через сколько лет клиент будет иметь на счету сумму не менее Z рублей. На вход программе идут два числа - Y, Z. 
Внимание: в процессе выполнения программы изменяйте переменную Y.

In [3]:
Z=int(input('Введите сумму, которую вы хотите иметь на счету:'))  
Y=int(input('Введите сумму, которую вы кладете на счет:'))        
year=0
yearly_multiplier=1.1
while Y<Z:
    Y=Y*yearly_multiplier
    year=year+1
else:
    print('Вы будете иметь эту сумму на счету через', year, 'лет')

Введите сумму, которую вы хотите иметь на счету:1000000
Введите сумму, которую вы кладете на счет:170000
Вы будете иметь эту сумму на счету через 19 лет


Задание 2 
Пассажирский лифт имеет ограничение на перевозку - не более 400 кг единовременно. Пусть на вход программе поступают веса (weight) входящих в лифт людей, а в случае, когда возникает перевес, программа выдаёт предупреждение: "Перевес N кг" (при перевесе в 15 кг "Перевес 15 кг") и указывает, насколько текущий вес пассажиров превзошёл максимум. 
Для проверки используйте weight = 77 и не меняйте его по ходу цикла.

In [5]:
total_weight=0
while total_weight<400:
    weight=int(input('Введите вес зашедшего: '))
    total_weight=total_weight+weight
else:
    print('Перевес', total_weight-400, 'кг')

Введите вес зашедшего: 100
Введите вес зашедшего: 100
Введите вес зашедшего: 100
Введите вес зашедшего: 114
Перевес 14 кг


Задание 3 
Вы играете в компьютерную игру, дошли до схватки с финальным боссом, но вот беда! Ваш компьютер заглючил, и вы не можете управлять персонажем в игре. Босс вас аттакует и каждую секунду наносит вам один удар, который отнимает у вас 80 единиц здоровья. На схватку с боссом вы вышли с 500 единицами здоровья. Создайте цикл, позволяющий понять, через сколько секунд босс вас убьет. 
Для проверки используйте переменную current_health для сохранения текущего уровня здоровья и изменяйте её по ходу цикла, и переменную attack = 80 для хранения значения атаки Босса.

In [6]:
damage=0
time=0
while damage<500:
    damage=damage+80
    time=time+1
else:
    print('через', time)

через 7


In [7]:
current_health=500
attack=80
time=0
while current_health>0:
    current_health=current_health-attack
    time=time+1
else:
    print('через', time)

через 7


Задание 4 
На банковской карте находится 10 000 рублей. Её хозяин имел привычку записывать все свои расходы в специальный дневник, но, немного научившись программировать, он решил автоматизировать процесс. Он написал программу, в которую вводит свои расходы (spent) (в рублях), а программа показывает баланс (balance) (При остатке 3500 показывается сообщение 3500). Как выглядел код этой программы? (Когда списание переходит через 0 в балансе, программа выводит "Слишком большие расходы") 
Для проверки используйте переменную spent = 2800 и не меняйте её по ходу цикла.

In [13]:
balance=10000
while balance>0:
    spent=int(input('Введите расходы:'))
    balance=balance-spent
    if balance<0:
            continue
    print(balance)
    
else:
    print('Слишком большие расходы')

Введите расходы:6000
4000
Введите расходы:5000
Слишком большие расходы


In [ ]:
balance=10000             #принято проверкой
spent=2800
while balance>spent:
    balance=balance-spent
    print(balance)
else:
    print('Слишком большие расходы')

Задание 5 
Представьте, что перед вами большая цистерна объёмом 1000 литров. Её необходимо наполнить полностью водой и при этом не допустить того, чтобы вода перелилась через край. У вас есть робот, который доставляет в цистерну сначала 5 литров, а затем каждый раз на 5 литров больше, чем в предыдущий. Изначально в цистерне нет воды. Выведите на экран через сколько заходов робота вода польётся через край?
Для проверки используйте переменную volume, равную свободному объёму в цистерне, и меняйте её по ходу цикла (на последнем шаге она должна быть меньше нуля).

In [14]:
volume=1000
add=5
time=0
while volume>0:
    volume=volume-add
    add=add+5
    time=time+1
else:
    print(time)

20
